File: go/mechanisms/svm/upto/facilitator/signer.go
package facilitator

import (
	"context"
	"fmt"

	solana "github.com/gagliardetto/solana-go"
	"github.com/gagliardetto/solana-go/rpc"

	"github.com/x402-foundation/x402/go/v2/mechanisms/svm"
	"github.com/x402-foundation/x402/go/v2/mechanisms/svm/paymentchannels"
)

// UptoFacilitatorSigner is FacilitatorSvmSigner narrowed to the read and
// simulate RPC the `upto` facilitator requires at runtime. Exact-only signers
// omit the extra methods. GetProgramAccounts is optional and only needed for
// RentCleanupManager.Discover(); assertProgramAccountsSigner checks it.
type UptoFacilitatorSigner interface {
	svm.FacilitatorSvmSigner

	GetAccountInfo(
		ctx context.Context,
		account solana.PublicKey,
		network string,
		opts *rpc.GetAccountInfoOpts,
	) (*rpc.GetAccountInfoResult, error)
	GetLatestBlockhash(ctx context.Context, network string) (solana.Hash, uint64, error)
	GetSlot(ctx context.Context, network string, commitment rpc.CommitmentType) (uint64, error)
	SimulateTransactionWithOpts(
		ctx context.Context,
		tx *solana.Transaction,
		network string,
		opts *rpc.SimulateTransactionOpts,
	) error
}

// assertUptoFacilitatorSigner validates that signer exposes every RPC cap the
// `upto` facilitator needs at construction. Panics with a clear message when a
// required method is missing.
func assertUptoFacilitatorSigner(signer svm.FacilitatorSvmSigner, label string) UptoFacilitatorSigner {
	type accountInfoGetter interface {
		GetAccountInfo(context.Context, solana.PublicKey, string, *rpc.GetAccountInfoOpts) (*rpc.GetAccountInfoResult, error)
	}
	type blockhashGetter interface {
		GetLatestBlockhash(context.Context, string) (solana.Hash, uint64, error)
	}
	type slotGetter interface {
		GetSlot(context.Context, string, rpc.CommitmentType) (uint64, error)
	}
	type simWithOpts interface {
		SimulateTransactionWithOpts(context.Context, *solana.Transaction, string, *rpc.SimulateTransactionOpts) error
	}

	if _, ok := signer.(accountInfoGetter); !ok {
		panic(fmt.Sprintf("%s requires GetAccountInfo on the signer", label))
	}
	if _, ok := signer.(blockhashGetter); !ok {
		panic(fmt.Sprintf("%s requires GetLatestBlockhash on the signer", label))
	}
	if _, ok := signer.(slotGetter); !ok {
		panic(fmt.Sprintf("%s requires GetSlot on the signer", label))
	}
	if _, ok := signer.(simWithOpts); !ok {
		panic(fmt.Sprintf("%s requires SimulateTransactionWithOpts on the signer", label))
	}
	return signer.(UptoFacilitatorSigner)
}

// programAccountsGetter is the optional discovery sweep cap. It is not part of
// UptoFacilitatorSigner so NewUptoSvmScheme can construct without it.
type programAccountsGetter interface {
	GetProgramAccounts(
		context.Context,
		string,
		solana.PublicKey,
		*rpc.GetProgramAccountsOpts,
	) (rpc.GetProgramAccountsResult, error)
}

// assertProgramAccountsSigner validates GetProgramAccounts for discovery sweeps.
func assertProgramAccountsSigner(signer svm.FacilitatorSvmSigner, label string) programAccountsGetter {
	gpa, ok := signer.(programAccountsGetter)
	if !ok {
		panic(fmt.Sprintf("%s requires GetProgramAccounts on the signer", label))
	}
	return gpa
}

type programAccountsQuerier struct {
	signer  programAccountsGetter
	network string
}

func (q programAccountsQuerier) GetProgramAccounts(
	ctx context.Context,
	opts *rpc.GetProgramAccountsOpts,
) (rpc.GetProgramAccountsResult, error) {
	return q.signer.GetProgramAccounts(ctx, q.network, paymentchannels.ProgramID, opts)
}

File: typescript/packages/mechanisms/svm/src/upto/facilitator/signer.ts
import type { Address } from "@solana/kit";

import type { FacilitatorSigningCapabilities, FacilitatorSvmSigner } from "../../signer";

/**
 * {@link FacilitatorSvmSigner} narrowed to the optional RPC caps the `upto`
 * facilitator requires at runtime. Exact-only signers omit the read methods.
 */
export type UptoFacilitatorSigner = FacilitatorSvmSigner & {
  getAccountInfo: NonNullable<FacilitatorSvmSigner["getAccountInfo"]>;
  getLatestBlockhash: NonNullable<FacilitatorSvmSigner["getLatestBlockhash"]>;
  getSlot: NonNullable<FacilitatorSvmSigner["getSlot"]>;
  getSigner(feePayer: Address): FacilitatorSigningCapabilities;
};

/**
 * Assert a facilitator signer exposes every optional cap `upto` needs.
 *
 * @param signer - Facilitator signer to validate
 * @param label - Scheme or component name for error messages
 * @throws Error when a required capability is missing
 */
export function assertUptoFacilitatorSigner(
  signer: FacilitatorSvmSigner,
  label = "UptoSvmScheme",
): asserts signer is UptoFacilitatorSigner {
  if (typeof signer.getSigner !== "function") {
    throw new Error(`${label} requires getSigner on the signer.`);
  }
  if (typeof signer.getAccountInfo !== "function") {
    throw new Error(`${label} requires getAccountInfo on the signer.`);
  }
  if (typeof signer.getLatestBlockhash !== "function") {
    throw new Error(`${label} requires getLatestBlockhash on the signer.`);
  }
  if (typeof signer.getSlot !== "function") {
    throw new Error(`${label} requires getSlot on the signer.`);
  }
}
